borg/helpers.py

```diff
@@ -1080,3 +1082,111 @@ def log_multi(*msgs, level=logging.INFO):
         lines.extend(msg.splitlines())
     for line in lines:
         logger.log(level, line)
+
+
+class ItemFormatter:
+    FIXED_KEYS = {
+        # Formatting aids
+        'LF': '\n',
+        'SPACE': ' ',
+        'TAB': '\t',
+        'CR': '\r',
+        'NUL': '\0',
+        'NEWLINE': os.linesep,
+    }
+
+    @classmethod
+    def available_keys(cls):
+        class FakeArchive:
+            name = ""
+
+        formatter = cls(FakeArchive, "")
+        keys = []
+        keys.extend(formatter.call_keys.keys())
+        keys.extend(formatter.get_item_data({
+            b'mode': 0, b'path': '', b'user': '', b'group': '', b'mtime': 0,
+            b'uid': 0, b'gid': 0,
+        }).keys())
+        keys.sort()
+        keys.sort(key=str.isupper)
+        return keys
+
+    def __init__(self, archive, format):
+        self.archive = archive
+        self.format = format
+        self.format_keys = {f[1] for f in Formatter().parse(format)}
+        self.call_keys = {
+            'size': self.calculate_size,
+            'isomtime': partial(self.format_time, b'mtime'),
+            'isoctime': partial(self.format_time, b'ctime'),
+            'isoatime': partial(self.format_time, b'atime'),
+            'mtime': partial(self.time, b'mtime'),
+            'ctime': partial(self.time, b'ctime'),
+            'atime': partial(self.time, b'atime'),
+        }
+        for hash_function in hashlib.algorithms_guaranteed:
+            self.call_keys[hash_function] = partial(self.hash_item, hash_function)
+        self.keys = set(self.call_keys) & self.format_keys
```

Comment:
Intersection is what we want:  call_keys are alle keys calculated from callables, format_keys are those used by the format string. The intersecting set are those keys used by the format string which are contained in call_keys and need to be sourced from there.

Variable naming should be improved :D